File: Programming Fundamentials/Objects and Classes/Exercises/Book Library Modification.cs
using System;
using System.Linq;
using System.Text;
using System.Collections.Generic;
using System.Globalization;
using static System.Console;

class Book
{
    public string Title             { set; get; }
    public string Publisher         { set; get; }
    public DateTime ReleaseDate     { set; get; }
    string ISBN                     { set; get; }
    public decimal Price            { set; get; }

    public Book(string title, string publisher, string releaseDate, string ISBN, string price)
    {
        Title = title;
        Publisher = publisher;
        ReleaseDate = DateTime.ParseExact(releaseDate, "d.MM.yyyy", CultureInfo.InvariantCulture);
        this.ISBN = ISBN;
        Price = decimal.Parse(price);
    }
}

class Library
{
    Dictionary<string, List<Book>> authorBooks        = new Dictionary<string, List<Book>>();

    public void AddAuthor(string name, Book book)     => authorBooks[name] = new List<Book> { book };

    public void AddBook(string authorName, Book book) => authorBooks[authorName].Add(book);

    public bool HasAuthor(string name)                => authorBooks.ContainsKey(name);

    public string ListAuthorSalery()
    {
        var authorSalery = authorBooks.ToDictionary(k => k.Key, v => v.Value.Sum(a => a.Price));
        authorSalery = authorSalery.OrderByDescending(p => p.Value).ThenBy(a => a.Key)
            .ToDictionary(k => k.Key, v => v.Value);

        var sB = new StringBuilder();
        foreach (var author in authorSalery)
            sB.AppendLine(author.Key + " -> " + author.Value.ToString("f2"));

        return sB.ToString();
    }

    public string ListBooksAfterDate(string day, string month, string year)
    {
        string formatedDate = $"{day}.{month}.{year}";
        var date = DateTime.ParseExact(formatedDate, "d.MM.yyyy", CultureInfo.InvariantCulture);

        var allBooksList = authorBooks.Values.ToList();
        var booksAfterCertainDate = new List<Book>();

        foreach(var bookList in allBooksList)
            foreach(var book in bookList)
                if (book.ReleaseDate.Date > date.Date)
                    booksAfterCertainDate.Add(book);

        booksAfterCertainDate = booksAfterCertainDate.OrderBy(b=>b.ReleaseDate.Date).ThenBy(b=>b.Title).ToList();

        var bookData = new StringBuilder();
        foreach (var book in booksAfterCertainDate)
            bookData.AppendLine(book.Title + " -> " + book.ReleaseDate.Date.ToString("d.MM.yyyy"));
        return bookData.ToString();
    }
}

class Program
{
    static void Main()
    {
        var lib = new Library();

        for (int i = int.Parse(ReadLine()); i > 0; i--)
        {
            var input = ReadLine().Split().ToArray();
            string bookName    = input[0];
            string bookAuthor  = input[1];
            string publisher   = input[2];
            string releaseDate = input[3];
            string ISBN        = input[4];
            string price       = input[5];

            if (lib.HasAuthor(bookAuthor))
                lib.AddBook(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
            else
                lib.AddAuthor(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
        }
        string[] afterDate = ReadLine().Split('.');

        Write(lib.ListBooksAfterDate(afterDate[0],afterDate[1],afterDate[2]));
    }
}

File: Software Technologies/Java/Exercises/Book Library Modification.java
package com.company;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.lang.System.in;

public class Main {
    static Scanner Input = new Scanner(in);

    public static void main(String[] args) {
        Library lib = new Library();

        for (int i = Integer.parseInt(Input.nextLine()); i > 0; i--)
        {
            String[] input     = Input.nextLine().split(" ");
            String bookName    = input[0];
            String bookAuthor  = input[1];
            String publisher   = input[2];
            String releaseDate = input[3];
            String ISBN        = input[4];
            String price       = input[5];

            if (lib.HasAuthor(bookAuthor))
                lib.AddBook(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
            else
                lib.AddAuthor(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
        }
        String afterDate = Input.nextLine();
        System.out.println(lib.ListBooksAfterDate(afterDate));
    }
}

class Book {
    SimpleDateFormat dateFormat = new SimpleDateFormat("d.MM.yyyy");

    public String Title;
    public String Publisher;
    public Calendar ReleaseDate = Calendar.getInstance();
    String ISBN;
    public Double Price;

    public Book(String title, String publisher, String releaseDate, String ISBN, String price)
    {
        Title = title;
        Publisher = publisher;
        try { ReleaseDate.setTime(dateFormat.parse(releaseDate)); } catch (ParseException ex){}
        this.ISBN = ISBN;
        Price = Double.parseDouble(price);
    }
}

class Author{
    public String Name;
    public ArrayList<Book> Books = new ArrayList<>();

    public  Author(String name, Book book){
        Name=name;
        Books.add(book);
    }

    public Double getSallery() {
        Double sum = 0.00;
        for(Book book:Books)
            sum+=book.Price;
        return sum;
    }
}

class Library
{
    private List<Author> authors = new ArrayList<>();

    public void AddAuthor(String name, Book book){
        authors.add(new Author(name,book));
    }

    public void AddBook(String authorName, Book book) {
        for(Author author: authors)
            if(author.Name.compareTo(authorName)==0)
                author.Books.add(book);
    }

    public Boolean HasAuthor(String name) {
        for(Author author:authors)
            if(author.Name.compareTo(name)==0)
                return true;
        return false;
    }

    public String ListBooksAfterDate(String date){
        try{
            SimpleDateFormat df = new SimpleDateFormat("d.MM.yyyy");

            Date dateToCompare = df.parse(date);
            List<Book> books = getAllBooks(authors);
            Comparator<Book> compareByDateThenByTitle = Comparator.comparing(b->b.ReleaseDate.getTime());
            compareByDateThenByTitle = compareByDateThenByTitle.thenComparing(b->b.Title);

            List<Book> booksToList = books
                    .stream()
                    .sorted(compareByDateThenByTitle)
                    .filter(b->b.ReleaseDate.getTime().compareTo(dateToCompare)>0)
                    .collect(Collectors.toList());

            StringBuffer list = new StringBuffer();
            for(Book book:booksToList)
                list.append(book.Title+" -> "+df.format(book.ReleaseDate.getTime())+"\n");  //display the datetime in wanted format
            return list.toString();
        }catch (ParseException ex){}
        return null;
    }

    private List<Book> getAllBooks(List<Author> authors) {
        List<Book> outList = new ArrayList<>();
        for(Author author:authors)
            outList.addAll(author.Books);
        return outList;
    }
}
